src/components/tl/cuda/bcast/bcast_linear.c

```diff
@@ -0,0 +1,303 @@
+            sbuf = PTR_OFFSET(TASK_SCRATCH(task, task->bcast_linear.root),
+                              task->bcast_linear.step % 2 * chunk_size);
+            st   = ecopy(dbuf, sbuf, chunk_size, exec,
+                       &task->bcast_linear.exec_task);
+            if (st != UCC_OK) {
+                ucc_error("failed to post ecopy task at client");
+                task->super.status = st;
+                return;
+            }
+            task->bcast_linear.stage = STAGE_CLIENT_COPY_WAIT;
+        case STAGE_CLIENT_COPY_WAIT:
+            etask = task->bcast_linear.exec_task;
+            if (etask) {
+                st = ucc_ee_executor_task_test(etask);
+                if (st == UCC_OK) {
+                    ucc_ee_executor_task_finalize(etask);
+                    task->bcast_linear.exec_task = NULL;
+                    ++task->bcast_linear.step;
+                    set_rank_step(task, trank, task->bcast_linear.step, 0);
+                    if (task->bcast_linear.step <
+                        task->bcast_linear.num_steps) {
+                        task->bcast_linear.stage = STAGE_WAIT_ROOT;
+                        return;
+                    } else {
+                        task->bcast_linear.stage = STAGE_DONE;
+                    }
+                } else {
+                    return;
+                }
+            } else {
+                return;
+            }
+        case STAGE_DONE:
+            task->super.status = UCC_OK;
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+ucc_status_t ucc_tl_cuda_bcast_linear_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_cuda_task_t *task = ucc_derived_of(coll_task, ucc_tl_cuda_task_t);
+    ucc_tl_cuda_team_t *team = TASK_TEAM(task);
+    ucc_coll_args_t *   args = &TASK_ARGS(task);
+    ucc_datatype_t      dt   = task->bcast_linear.dt;
+    size_t              half_scratch_size = get_raw_scratch_size(team) / 2;
+
+    task->bcast_linear.stage = STAGE_SYNC;
+
+    task->bcast_linear.size = ucc_dt_size(dt) * args->src.info.count;
+    task->bcast_linear.num_steps =
+        ucc_div_round_up(task->bcast_linear.size, half_scratch_size);
+
+    ucc_debug("bcast linear dt: %s, buffer size: %ld, num_steps: %d",
+              ucc_datatype_str(dt), task->bcast_linear.size,
+              task->bcast_linear.num_steps);
+
+    task->bcast_linear.sbuf = args->src.info.buffer;
+    task->bcast_linear.step = 0;
+
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(team)->pq, &task->super);
+}
+
+ucc_status_t ucc_tl_cuda_bcast_linear_init(ucc_base_coll_args_t *coll_args,
+                                           ucc_base_team_t *     tl_team,
+                                           ucc_coll_task_t **    task_p)
+{
+    ucc_tl_cuda_team_t *team = ucc_derived_of(tl_team, ucc_tl_cuda_team_t);
+    ucc_tl_cuda_task_t *task;
+    ucc_status_t        status;
+
+    if (ucc_unlikely(!ucc_tl_cuda_team_topo_is_fully_conntected(team->topo) ||
```

Comment:
I understand it is not related to the pr though... but I'm pointing to the typo anyway
```suggestion
    if (ucc_unlikely(!ucc_tl_cuda_team_topo_is_fully_connected(team->topo) ||
```

Reply:
fixed typo